### PR TITLE
Bug 1202376 - [browser] Implement Bookmark folder navigation in TV

### PR DESCRIPTION
**Pull request by danhuang1202**
- add smart_list.js for display bookmark
- bind event handler with awesomescreen.js
- fix js lint error in awesomescreen.js

### CODE REVIEW

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=187284f6a31dd232407326f9cd92d2d4a1edc901

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=6857ddd7455b3ccfcf4595f921c71ff69d856cc8

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=373fcc76b8536d6b9b05e5c885600a1337cf5626

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=ead255bea33c02ccf21ca3d571ff1127ac23f4a5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=25e58db3fc3985e941c758e3c8757511a7ce9787

File: tv_apps/browser/js/smart_list.js [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+			typeof options.handleListMouseOut === 'function' ?
+			options.handleListMouseOut : null;
+
+		this.handleListMouseOver = options.handleListMouseOver &&
+			typeof options.handleListMouseOver === 'function' ?
+			options.handleListMouseOver : null;
+	}
+
+	SmartList.prototype = {
+		init: function(options){
```
Comment (begeeben) [inline tv_apps/browser/js/smart_list.js]: Missing ```constructor: SmartList,```. Although this doesn't affect the feature, changing prototype like this removes its constructor property.

File: tv_apps/browser/style/smart_list.css [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  background-color: rgba(255,255,255,0.2);
+}
+
+.smart-list .list-item[data-type="button"]:focus .icon-box .icon-img:before {
+  color: #fff;
+}
+
+.smart-list .list-item[data-type="button"]:focus .text-box .title {
+  color: #fff;
+}
```
Comment (begeeben) [inline tv_apps/browser/style/smart_list.css]: Missing a newline.

File: tv_apps/browser/js/smart_list.js [hunk 3]
```diff
@@ -0,0 +1,346 @@
...
+			var i = 0,
+					len = SMART_LIST_EVENTS.length;
+			for(; i < len; i++) {
+				dstEl[SMART_LIST_EVENTS[i]] = srcEl[SMART_LIST_EVENTS[i]];
+			}
+		}
+	};
+
+	exports.SmartList = SmartList;
+})(window);
```
Comment (begeeben) [inline tv_apps/browser/js/smart_list.js]: Missing a newline.

File: tv_apps/browser/js/smart_list.js [hunk 4]
```diff
@@ -0,0 +1,346 @@
...
+
+				if(this.onRender) {
+					this.onRender();
+				}
+			}
+		},
+
+		cleanList: function(){
+			if(this.container) {
+				this.container.innerHTML = '';
```
Comment (begeeben) [inline tv_apps/browser/js/smart_list.js]: Please replace ```innerHTML``` with
```
// Removing all children from an element
var element = document.getElementById("top");
while (element.firstChild) {
  element.removeChild(element.firstChild);
}
```.

**Comment by begeeben** [comment]
Please replace
```
      if(( Browser.currentInfo.url != null ) &&
        ( Browser.currentInfo.url !== '' )) {
```
to
```
      if(Browser.currentInfo.url) {
```
in awesomescreen.js.

**Comment by begeeben** [comment]
Ideally we would like to leave sync status in ```Settings.firefoxAccount```. Please change ```bookmarkFromFirefoxAccount``` to ```Settings.firefoxAccount.syncBookmark``` in awesomescreen.js, thanks!

**Comment by begeeben** [comment]
At ln.642 in awesomescreen.js ```this.awesomescreen.classList.add('awesomescreen-bookmark');```, I would suggest we move this line out of the if-else clause and remove the same code at ln.3697. Since SmartList.render has no async callback.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c745abaa3fde1456bcf4ee254e8bbda59e954fc5

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=2abc589bab5f59698c63e428ca5d00ce5b942166

File: tv_apps/browser/index.html [hunk 5]
```diff
@@ -195,7 +199,7 @@
                     <g>
                       <g>
                         <polygon class="bookmark-button-path" fill="#00caf2" points="43.116,22.382 31.878,20.904 27,10.673 22.122,20.904 10.884,22.382 19.107,30.183 17.04,41.327
-                        27,35.917 36.96,41.327 34.893,30.183 			"></polygon>
+                        27,35.917 36.96,41.327 34.893,30.183                   "></polygon>
```
Comment (begeeben) [inline tv_apps/browser/index.html]: I'm curious about what's the reason for this change?

File: tv_apps/browser/style/bookmarkList.css [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  height: 100%;
+  padding: 2rem;
+}
+
+.bookmark-list .list-item .text-box {
+  display: inline-flex;
+  width: 111.5rem;
+  height: 100%;
+  padding-top: 15px;
+  padding-bottom: 15px;
```
Comment (begeeben) [inline tv_apps/browser/style/bookmarkList.css]: Please change the unit to ```rem```.

File: tv_apps/browser/bower.json [hunk 7]
```diff
@@ -16,5 +16,6 @@
   "dependencies": {
     "evt": "gaia-components/evt#~0.4.0",
     "smart-button": "gaia-components/smart-button"
+    "smart-icons": "smart-components/smart-icons#~0.0.2"
```
Comment (begeeben) [inline tv_apps/browser/bower.json]: Missing a comma at line 18.

**Comment by begeeben** [comment]
https://github.com/danhuang1202/gaia/blob/1202376_tv_browser_bookmark_navigation/tv_apps/browser/js/awesomescreen.js#L210

Please change the name from ```enable```/```disable``` to ```show```/```hide```.

**Comment by begeeben** [comment]
https://github.com/danhuang1202/gaia/blob/1202376_tv_browser_bookmark_navigation/tv_apps/browser/js/awesomescreen.js#L658

Spelling error.

**Comment by begeeben** [comment]
https://github.com/danhuang1202/gaia/blob/1202376_tv_browser_bookmark_navigation/tv_apps/browser/js/awesomescreen.js#L2494

```if ( Browser.info[tabIds[i]].url ) {```

File: tv_apps/browser/js/bookmarkList.js [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+  }
+
+  BookmarkList.prototype = {
+    constructor: BookmarkList,
+
+    /**
+     * Open bookmark list.
+     * @param  {Array} bookmarks - Array of bookmark data.
+     */
+    open: function(bookmarks) {
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Looks like the ```bookmarks``` variable here hasn't been used in your patch.

File: tv_apps/browser/js/bookmarkList.js [hunk 9]
```diff
@@ -0,0 +1,839 @@
...
+      var eventDetail = {
+        startAt: 0,
+        number: MAX_VISIBLE_ITEM + 7,
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Will it be better with ```MAX_VISIBLE_ITEM * 2``` ?

File: tv_apps/browser/js/bookmarkList.js [hunk 10]
```diff
@@ -0,0 +1,839 @@
...
+    this.navState = '';
+
+    /**
+     * Save user browser history.
+     * @type {Array}
+     */
+    this.navHistory = [];
+
+    /**
+     * list items which in dom tree and been initialed but render with data
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Spelling error? ```initialized```?

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=95f662fd7769691efc234ac2026c0d178f178ada

File: tv_apps/browser/js/bookmarkList.js [hunk 11]
```diff
@@ -0,0 +1,839 @@
...
+       *  if current render folder is not ROOT_FOLDER,
+       *  render back button element first
+       */
+      if(this.navState != ROOT_FOLDER) {
+        bookmarks.unshift(this.generateBackButtonData());
+      }
+
+      for(; i < length; i++) {
+        /*
+         *  if there are reseted list item, user these to render data,
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Spelling error? ```use```?

File: tv_apps/browser/js/bookmarkList.js [hunk 12]
```diff
@@ -0,0 +1,839 @@
...
+      itemEl.
+        addEventListener('keydown', this.handleItemKeyDown.bind(this));
+      itemEl.
+        addEventListener('keyup', this.handleItemKeyUp.bind(this));
+      itemEl.
+        addEventListener('mouseup', this.handleItemMouseUp.bind(this));
+      itemEl.
+        addEventListener('mouseover', this.handleItemMouseOver.bind(this));
+      itemEl.
+        addEventListener('mouseout', this.handleItemMouseOut.bind(this));
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Adding too many event listeners will consume memory. Please add ```// XXX: delegate these events later``` and file a follow up bug for it.

File: tv_apps/browser/js/bookmarkList.js [hunk 13]
```diff
@@ -0,0 +1,839 @@
...
+        } else {
+          this.dispatchLoadDataByIndex(
+            bufferIndex,
+            bufferIndex,
+            this.navState
+          );
+        }
+      }
+
+      // preload the item before current focus item
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: ```after```

File: tv_apps/browser/js/bookmarkList.js [hunk 14]
```diff
@@ -0,0 +1,839 @@
...
+      titleEl.textContent = '';
+      uriEl.textContent = '';
+    },
+
+    /**
+     * Preload the item before and after the view item.
+     * @param  {Number} index - current focus index.
+     */
+    loadItemBuffer: function(index) {
+      var bufferIndex = 0;
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: I'm a bit worried of 2 things, 1) what if the data querying takes longer than expected. 2) preloading data in both ways seems not ideal. We can go with this now.

File: tv_apps/browser/js/bookmarkList.js [hunk 15]
```diff
@@ -0,0 +1,839 @@
...
+          type = targetEl.getAttribute('data-type'),
+          event = null,
+          eventDetail = null;
+
+      switch(type) {
+        case 'folder':
+          var folderId = targetEl.getAttribute('data-folder');
+          eventDetail = {
+            startAt: 0,
+            number: MAX_VISIBLE_ITEM + 6,
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Is ```6``` has any special meaning here? Should we use ```MAX_VISIBLE_ITEM``` as well?

File: tv_apps/browser/js/bookmarkList.js [hunk 16]
```diff
@@ -0,0 +1,839 @@
...
+            folderId: folderId,
+            callback: (function(bookmarks) {
+              this.navHistory.push(folderId);
+              this.render(bookmarks);
+              if(this.listIndexEndAt != -1) {
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Please use ```!==```.

File: tv_apps/browser/js/bookmarkList.js [hunk 17]
```diff
@@ -0,0 +1,839 @@
...
+                var focusEl = this.listItemMap[0];
+                if(focusEl.getAttribute('data-type') == 'button') {
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Please use ```===```.

File: tv_apps/browser/js/bookmarkList.js [hunk 18]
```diff
@@ -0,0 +1,839 @@
...
+            this.dispatchDisplayWebsite(uriEL.textContent);
+            this.close();
+          }
+          break;
+        case 'button':
+          this.navHistory.pop();
+          this.navState = this.getCurNavHistory();
+          eventDetail = {
+            startAt: 0,
+            number: MAX_VISIBLE_ITEM + 6,
```
Comment (begeeben) [inline tv_apps/browser/js/bookmarkList.js]: Same as above.

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=845e5a5a2a808d5fb8bf77e88c529ec94dbc598a

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=a8858cfeb1b4ec54c824bc5020151ca262fd6753
